compiler/src/dotty/tools/dotc/core/Denotations.scala

```diff
@@ -433,72 +438,54 @@ object Denotations {
-          else if (isDoubleDef(sym1, sym2)) handleDoubleDef
-          else {
-            val sym =
-              if (preferSym(sym2, sym1)) sym2
-              else sym1
-            val jointInfo =
-              try infoMeet(info1, info2, sym1, sym2, safeIntersection)
-              catch {
-                case ex: MergeError =>
-                  // TODO: this picks one type over the other whereas it might be better
-                  // to return a MultiDenotation instead. But doing so would affect lots of
-                  // things, starting with the return type of this method.
-                  if (preferSym(sym2, sym1)) info2
-                  else if (preferSym(sym1, sym2)) info1
-                  else if (pre.widen.classSymbol.is(Scala2x) || migrateTo3)
-                    info1 // follow Scala2 linearization -
-                  // compare with way merge is performed in SymDenotation#computeMembersNamed
-                  else throw new MergeError(ex.sym1, ex.sym2, ex.tp1, ex.tp2, pre)
-              }
-            new JointRefDenotation(sym, jointInfo, denot1.validFor & denot2.validFor, pre)
-          }
-        }
-      }
+        def isHidden(sym: Symbol) = sym.exists && !sym.isAccessibleFrom(pre)
+        val hidden1 = isHidden(sym1)
+        val hidden2 = isHidden(sym2)
+        if hidden1 && !hidden2 then denot2
+        else if hidden2 && !hidden1 then denot1
+        else
+          // The score that determines which symbol to pick for the result denotation.
+          // A value > 0 means pick `sym1`, < 0 means pick `sym2`.
+          // A value of +/- 2 means pick one of the denotations as a tie-breaker
+          // if a common info does not exist.
+          val symScore: Int =
+            if !sym1.exists then -2
+            else if !sym2.exists then 2
+            else if sym1.is(Bridge) && !sym2.is(Bridge) then -2
+            else if sym2.is(Bridge) && !sym1.is(Bridge) then 2
+            else if !sym1.isAsConcrete(sym2) then -1
+            else if !sym2.isAsConcrete(sym1) then 1
+            else
+              val linScore = linearScore(sym1.owner, sym2.owner)
+              if linScore != 0 then linScore
+              else
+                val boundary1 = accessBoundary(sym1)
+                val boundary2 = accessBoundary(sym2)
+                if boundary1.isProperlyContainedIn(boundary2) then -1
+                else if boundary2.isProperlyContainedIn(boundary1) then 1
+                else if sym2.is(Method) && !sym1.is(Method) then -1
+                else if sym1.is(Method) && !sym2.is(Method) then 1
+                else 0
+
+          val matchLoosely = sym1.matchNullaryLoosely || sym2.matchNullaryLoosely
+
+          if info2.overrides(info1, matchLoosely, checkClassInfo = false)
+             && symScore <= 0
+          then denot2
+          else if info1.overrides(info2, matchLoosely, checkClassInfo = false)
+                  && symScore >= 0
```

Comment:
Looks like we can reorder the checks to avoid potentially expensive calls to `overrides`:
```suggestion
          if symScore <= 0
             && info2.overrides(info1, matchLoosely, checkClassInfo = false)
          then denot2
          else if symScore >= 0
                  && info1.overrides(info2, matchLoosely, checkClassInfo = false)
```